system-tests/lib/docker.ts

```diff
@@ -0,0 +1,151 @@
+import type { SpawnerResult, Spawner } from './system-tests'
+import Docker from 'dockerode'
+import stream from 'stream'
+import EventEmitter from 'events'
+import path from 'path'
+import { promises as fs } from 'fs'
+import execa from 'execa'
+import Fixtures from './fixtures'
+import { nock } from './spec_helper'
+
+let docker: Docker | null = null
+
+const getDocker = () => {
+  return docker || (docker = new Docker())
+}
+
+const log = (...args) => {
+  console.error('🐋', ...args)
+}
+
+class DockerProcess extends EventEmitter implements SpawnerResult {
+  stdout = new stream.PassThrough()
+  stderr = new stream.PassThrough()
+
+  constructor (private dockerImage: string) {
+    super()
+  }
+
+  pull () {
+    return new Promise<void>((resolve, reject) => {
+      log('Pulling image', this.dockerImage)
+      getDocker().pull(this.dockerImage, null, (err, stream) => {
+        if (err) return reject(err)
+
+        const onFinished = (err) => {
+          log('Pull complete', { err })
+          if (err) return reject(err)
+
+          resolve()
+        }
+
+        const onProgress = (event) => {
+          log('Pull progress', JSON.stringify(event))
+        }
+
+        docker.modem.followProgress(stream, onFinished, onProgress)
+      }, null)
+    })
+  }
+
+  run (opts: {
+    cmd: string
+    args: string[]
+    env: any
+  }) {
+    const containerCreateEnv = []
+
+    for (const k in opts.env) {
+      // skip problematic env vars that we don't wanna preserve from `process.env`
+      if (['DISPLAY', 'USER', 'HOME', 'USERNAME', 'PATH'].includes(k)) continue
+
+      containerCreateEnv.push([k, opts.env[k]].join('='))
+    }
```

Comment:
I tend to prefer something like this, but it's your call:

```suggestion
    const containerCreateEnv = Object.entries(opts.env).filter(([key]) => !['DISPLAY', 'USER', 'HOME', 'USERNAME', 'PATH'].includes(key)).map((entry) => {
      return entry.join('=')
    })
```

Reply:
TIL about `Object.entries`. I only knew of `.keys` and `.values`. I think I'll keep it as-is since it's the clearest to me, but I'll keep this construct in mind.